File: py_payment_api/app/services/forwarder/websocket_forwarder.py
from typing import Dict, Any, Generic, TypeVar, Optional, final
from enum import StrEnum
import asyncio
import websockets
from .abstract.single_forwarder import SingleForwarder
from ...services import PurchaseStatus
from websockets.client import WebSocketClientProtocol

@final
class WebSocketForwarder(SingleForwarder):
    """Forwarder that sends data via WebSocket."""
    
    def __init__(self, url: str = "ws://localhost:8765") -> None:
        """Initialize the forwarder.
        
        Parameters
        ----------
        service : ItemCollectionService
            The service to use for status updates
        url : str, optional
            The WebSocket URL to connect to, by default "ws://localhost:8765"
        """
        super().__init__(PurchaseStatus.SENT_TO_WEBSOCKET)
        self._url = url
        self._websocket : Optional[WebSocketClientProtocol] = None

    async def connect(self) -> None:
        """Connect to the WebSocket server."""
        self._websocket = await websockets.connect(self._url)

    @final
    def _on_forward_event(self, data: Dict[str, Any]) -> None:
        """Forward event data via WebSocket.
        
        Parameters
        ----------
        data : Dict[str, Any]
            The data to forward
        """
        # Create event loop if it doesn't exist
        try:
            loop = asyncio.get_event_loop()
        except RuntimeError:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
        
        # Connect if not connected
        if not self._websocket:
            loop.run_until_complete(self.connect())
        
        # Send data
        if self._websocket:
            loop.run_until_complete(self._websocket.send(str(data)))

File: tsPaymentApi/src/services/forwarder/WebSocketForwarder.ts
import WebSocket from 'ws';
import { SingleForwarder } from '../../../../../../../tsPaymentApi/src/services/forwarder/Forwarder';
import { PurchaseStatus } from '../types/PurchaseStatus';

export class WebSocketForwarder extends SingleForwarder {
    private websocket: WebSocket | null = null;
    private readonly url: string;

    constructor(url: string = 'ws://localhost:8765') {
        super(PurchaseStatus.SENT_TO_WEBSOCKET);
        this.url = url;
    }

    private async connect(): Promise<void> {
        return new Promise((resolve, reject) => {
            this.websocket = new WebSocket(this.url);
            this.websocket.on('open', () => resolve());
            this.websocket.on('error', (error) => reject(error));
        });
    }

    protected async onForwardEvent(eventData: Record<string, any>): Promise<void> {
        if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN) {
            await this.connect();
        }

        if (this.websocket && this.websocket.readyState === WebSocket.OPEN) {
            return new Promise((resolve, reject) => {
                this.websocket!.send(JSON.stringify(eventData), (error) => {
                    if (error) {
                        reject(error);
                    } else {
                        resolve();
                    }
                });
            });
        } else {
            throw new Error('Failed to establish WebSocket connection');
        }
    }
} 